### PR TITLE
[FLINK-21947][csv] Support TIMESTAMP_LTZ type in CSV format

### PR DESCRIPTION
**Pull request by leonardBang**
## What is the purpose of the change

* This pull request properly support timestamp_ltz type in csv format

## Brief change log

 - use `SQL` dateformatter to format TIMESTAMP/TIMESTAMP_LZT date just like `Json` did
 - Use `LocalDateTime`for the  timestamp conversion  instead of `java.sql.Timestamp`

## Verifying this change

Add serialization/deserialization  tests for changes.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/TimeFormats.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+package org.apache.flink.formats.csv;
+
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.time.temporal.ChronoField;
+
+/** Time formats and timestamp formats respecting SQL specification. */
+class TimeFormats {
```
Comment (KurtYoung) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/TimeFormats.java:26]: We shouldn't copy this utility from `flink-json`. How about reuse this in a new module like `flink-format-common`?

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java [hunk 2]
```diff
@@ -316,7 +321,11 @@ private static RuntimeConverter createRuntimeConverter(
         } else if (info.equals(Types.LOCAL_TIME)) {
             return (node) -> Time.valueOf(node.asText()).toLocalTime();
         } else if (info.equals(Types.LOCAL_DATE_TIME)) {
-            return (node) -> Timestamp.valueOf(node.asText()).toLocalDateTime();
+            return (node) -> LocalDateTime.parse(node.asText(), SQL_TIMESTAMP_FORMAT);
+        } else if (info.equals(Types.INSTANT)) {
+            return (node) ->
+                    LocalDateTime.parse(node.asText(), SQL_TIMESTAMP_WITH_LOCAL_TIMEZONE_FORMAT)
+                            .toInstant(ZoneOffset.UTC);
```
Comment (KurtYoung) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDeserializationSchema.java:328]: I'm not sure we can use UTC here.

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/TimeFormats.java (hunk 1 above)
Comment (leonardBang) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/TimeFormats.java:26]: +1

**Comment by leonardBang** [comment]
@flinkbot run azure

**Comment by leonardBang** [comment]
The failed test is not related to this PR, it caused by download dependency error and reported in https://issues.apache.org/jira/browse/FLINK-19705

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java [hunk 3]
```diff
@@ -319,7 +323,12 @@ private static RuntimeConverter createRuntimeConverter(TypeInformation<?> info)
             return (csvMapper, container, obj) -> container.textNode(obj.toString());
         } else if (info.equals(Types.LOCAL_DATE_TIME)) {
             return (csvMapper, container, obj) ->
-                    container.textNode(DATE_TIME_FORMATTER.format((LocalDateTime) obj));
```
Comment (KurtYoung) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java:322]: delete useless DATE_TIME_FORMATTER

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/RowDataToCsvConverters.java [hunk 4]
```diff
@@ -268,8 +281,9 @@ private static JsonNode convertTime(int millisecond, ContainerNode<?> container)
         return container.textNode(ISO_LOCAL_TIME.format(time));
     }
 
-    private static JsonNode convertTimestamp(TimestampData timestamp, ContainerNode<?> container) {
-        return container.textNode(DATE_TIME_FORMATTER.format(timestamp.toLocalDateTime()));
```
Comment (KurtYoung) [inline flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/RowDataToCsvConverters.java:272]: delete useless DATE_TIME_FORMATTER

File: flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java (hunk 3 above)
Comment (KurtYoung) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java:322]: I noticed that this `SQL_TIMESTAMP_FORMAT` is not exactly the same with `DATE_TIME_FORMATTER`, please double check this won't cause any issue.

Comment (leonardBang) [reply flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowSerializationSchema.java:322]: I checked the change is safe, although `DATE_TIME_FORMATTER `  supports `2020-01-01 12:18` and `2020-01-01 12:18:00`, but the deserializer uses `Timestamp.valueOf(tiemstampString)`  which only supports `2020-01-01 12:18:00`, currently, and all serializer use `LocalDateTime` object which won't produce timestamp string like `2020-01-01 12:18`.

**Review by KurtYoung — APPROVED** [review]
LGTM

**Comment by KurtYoung** [comment]
@flinkbot run azure

**Comment by leonardBang** [comment]
@flinkbot run azure

**Comment by leonardBang** [comment]
@KurtYoung Thank for the merging, but the pipeline is still failed, I lookup it a little, we need to add for-common to `format-csv .jar` and  `format-csv.sql-jar.jar` , should I open a hotfix?

**Pull request closed without merging**